### DISCUSSION TITLE
Colormap generation: undefined variable error

### DISCUSSION THREAD

**Discussion by AugiteSoul** [ImageMagick/ImageMagick#4672] [Help]
Hello!
While trying to find ways to efficiently generate colormaps, I stumbled on [this](https://legacy.imagemagick.org/Usage/quantize/#remap) example in the documentation about generating a 332 colormap, which seemed like a good start for further experimentation.

After a few errors due to my lack of experience with Windows' command prompt, I ended up with this command:

`magick convert -size 16x16 xc: -channel R -fx "(i%8)/7" -channel G -fx "(j%8)/7" -channel B -fx "((i>>3&1)|(j>>2&2))/3" colormap_332.png`

However, I don't understand why does it return the following error:

`convert: undefined variable `(i)' @ error/fx.c/FxGetSymbol/1169.`

How would one get this command to work as expected? I'm aware that the "legacy" online documentation applies to ImageMagick version 6, but I am unaware of any similar documentation websites that would apply correctly to IM version 7.

By the way, here's the output of `magick -version`, if it is of any assistance:
`Version: ImageMagick 7.1.0-19 Q16-HDRI x64 2021-12-22 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2019 (192930138)`

Have a good day!

**Comment by fmw42** [comment]
What are you trying to do with the syntax "2&2"  That `&` is a binary AND.

**Reply by AugiteSoul** [reply]
Alright, after some troubleshooting and more of my lacking knowledge of Windows' batch file management, I pinpointed down the issue:
when running this command in a batch file, the `%` were completely ignored from the input, resulting in an incorrect colormap being generated:
![colormap_incorrect](https://user-images.githubusercontent.com/16467639/148280964-4b72add5-a532-4dfa-b5cd-df1a2b1a7f6c.png)

I apologize if I disturbed anyone by my lack of experience of ImageMagick, I'm barely getting around everything available.

**Comment by snibgo** [comment]
So, you now realise that in BAT scripts you need to double the percent signs: "%%" ?
